README.md
# COMP\_hack #

[![AGPL License](http://img.shields.io/badge/license-AGPL-brightgreen.svg)](https://opensource.org/licenses/AGPL-3.0)
[![Latest Release](https://img.shields.io/github/downloads/comphack/comp_hack/v4.12.1-wyrd/total.svg)](https://github.com/comphack/comp_hack/releases/tag/v4.12.1-wyrd)
[![Discord Chat](https://img.shields.io/discord/322024695266541579.svg)](http://discord.gg/9jXeKcJ)

[![Build Status](https://travis-ci.com/comphack/comp_hack.svg?branch=develop)](https://travis-ci.com/comphack/comp_hack)
[![Coverity Scan Build Status](https://scan.coverity.com/projects/9671/badge.svg)](https://scan.coverity.com/projects/comphack-comp_hack)
[![Coverage Status](https://coveralls.io/repos/github/comphack/comp_hack/badge.svg?branch=develop)](https://coveralls.io/github/comphack/comp_hack?branch=develop)
[![Documentation Status](https://readthedocs.org/projects/comp-hack/badge/?version=latest)](https://comp-hack.readthedocs.io/en/latest/?badge=latest)

## 真・女神転生IMAGINE Private Server ##

The is server software to revive an MMO that has been shutdown (SMT: IMAGINE). It's a complete re-implementation of the server from scratch and fully open source. The best place for documentation is the [Definitive Guide](https://comp-hack.readthedocs.io/en/latest/) so be sure to check it out.

### Building on Linux ###

You only need to build the project if you are on a Linux system that doesn't have a package (there is a [PPA](https://launchpad.net/~compomega/+archive/ubuntu/comphack) for Ubuntu-based systems) or you want to contribute. That being said, if you wish to contribute, Linux is the preferred build and run environment for the server. Of course you can build and develop with Visual Studio 2015 on Windows if that's your thing.

#### Dependencies ####

First thing you want to do is download some dependencies. Make sure you have GCC 5+ or Clang with C++14 support. Here is a command for Debian/Ubuntu based distros to pull in packages you may need:
```
sudo apt-get install build-essential cmake docbook-xsl doxygen texlive-font-utils xmlto libqt5webkit5-dev
```

#### Building ####

> Make sure to initialize and update the submodules before trying to build!

That should be all you need. Just build the project now:
```
mkdir build
cd build
cmake -DNO_WARNINGS=ON ..
make
```

See the [Definitive Guide](https://comp-hack.readthedocs.io/en/latest/chapters/hacking.html#build-system) for more information on the build system options and how to setup the server.

### Building on Windows ###

If you do not wish to contribute to the project, download from the [Releases](https://github.com/comphack/comp_hack/releases) section or download the nightly artifact off the [AppVeyor](https://ci.appveyor.com/project/compomega/comp-hack/history) page.

#### Required Dependencies ####

* [Visual Studio 2015](https://visualstudio.microsoft.com/vs/older-downloads/)
* [CMake](https://cmake.org)
* [Qt 5.7+](https://www.qt.io)

#### Optional Dependencies ####

* [Doxygen](http://www.doxygen.nl)
* [WiX Toolset](http://wixtoolset.org)

#### Building ####

> Make sure to initialize and update the submodules before trying to build!

Edit the _vsbuild_x86.bat_ and _vsbuild_x64.bat_ batch files to point to your install of Qt. Run the desired script and you should see a _build32_ or _build64_ folder. Inside the folder should be a comp_hack.sln solution file. Open the solution and build as normal.

See the [Definitive Guide](https://comp-hack.readthedocs.io/en/latest/chapters/hacking.html#build-system) for more information on the build system options and how to setup the server.


debian/changelog
comphack (4.12.2) xenial; urgency=low

  * Improved updater logging
  * Added the ability to select a soundtrack in the updater
  * Added the ability to configure the client DLL in the updater
  * Added feature to updater to notify it was used to start the game
  * Added detection of a high performance GPU for laptops (to the updater)
    * This feature requires an updated ImagineClient.exe and ImagineOption.exe which should be obtained elsewhere
  * Datastore has been added into the Windows installer and Linux packages
    * Updates can now be downloaded here: https://github.com/comphack/datastore
  * Added comp_translator - a cross-platform tool to automate building translated BinaryData files
  * Added generation of XML Schema Definitions (XSD files) for BinaryData
    * These can be used with comp_translator to validate the XML prior to conversion
    * These have been added to the installer
  * Added support for additional encodings to comp_bdpatch/comp_translator
    * See the usage of comp_bdpatch (or call AvailableEncodings in a comp_translator script) for the full list of encodings
    * Note that some encodings may not function properly in the client which expects a Multi-Byte Character Set (MBCS)
  * Look at the documentation of WorldSharedConfig for additional configuration options
  * Added acount dump capability to the test client
  * Crash when a team invite fails (#1230)
  * Skills with a DigitizeWeaponType restriction are applying the restriction on non-digitalize skills (#1226)
  * Issue where one stat point is lost every 5 levels (#1236)
    * Note that this includes a database migration (20200712_wyrd_stat_points.nut) to correct the stat points for all players. Migration will blindly apply additional points so make sure it is run only once.
  * Digi skills consider the normal weapon type as well (#1237)
  * Fixed crash when using the API which is mostly for casino games and account managment (#1252)
  * Crash on zone change when status effect was not properly removed (#1262)
  * Critical hits not occuring if target Critical Defense turned negative (#1266)
  * Items with a level restriction can not be unequipped when above the maximum level (#1263)
  * Tokusei aspect STATUS_INFLICT_ADJUST adjustments with a value of 0 do not affect all status effects (#1269)
  * Joining a party (with team members) after joining a Cathedral team kicks everyone involved out of the team (#1267)
    * Now they will only disband the team if somebody doesn't belong to it.
  * Enemies with 0% RATE_PC_TAKEN/RATE_DEMON_TAKEN are still affected by ailments on damage skills (#1246)
  * Demonic Compendium Summon Synchro length bonus does not apply; Mitama Demonic Compendium bonuses are inactive (#1245)
  * Pentalpha rankings and payouts are incorrect (#1281)
  * Damage-dealing status effects do not stop HP/MP Regeneration (#1244)
  * KNOCKBACK_STATUS_ADD should affect skills with a non-zero knockback modifier (#1256)
  * Demon friendship set to 0 if recruited via negotiation or event (#1292)
  * Status effects that are removed on skill usage and affect damage calculation are non-functional (#1261)
  * RATE_HEAL and RATE_HEAL_TAKEN are not applying if the user heals themselves (#1284)
  * Some skills used on spawns with a Category of ALLY do not process (#1257)
  * Entities with Status 901 (stealth) do not silently suppress targeting (#1283)
  * SpawnRestrictions on SpawnGroups were ignored when SpawnGroups were spawned by Squirrel scripting (#1300)
  * The server setting NRAStatusNull being false results in NRA not being checked at all for status ailments (#1306)
  * Incorrect formulae for multiple Item Mix extra items (#1309)
  * Add additional Flag values for the adjustRestrictions field of SkillData, for fixed HP/MP Costs (#1258)
  * Additional Sqrat binding for negotiation damage (#1239)
  * ActionUpdatePoints "isSet" property sets SOUL_POINTS to a value rather than giving a fixed amount (#1319)

 -- COMP Omega <compomega@tutanota.com>  Mon, 14 Dec 2020 11:04:29 -0700

comphack (4.12.1) xenial; urgency=low

  * QMP and pathing generation support
  * Added missing timed tokusei condition logic
  * Fixed invalid negotiation outcomes
  * Demon quests will "default to self" less
  * Incantation/cooldown/heal rate corrections
  * Tarot/soul availability after SI corrected

 -- COMP Omega <compomega@tutanota.com>  Sat, 27 Jun 2020 16:49:20 -0600

comphack (4.12.0.1) xenial; urgency=low

  * Corrected PPA build error.

 -- COMP Omega <compomega@tutanota.com>  Sat, 30 May 2020 14:11:55 -0600

comphack (4.12.0) xenial; urgency=low

  * Negotiation "progress" accuracy adjustments
  * Custom skill logic scripts
  * Configurable fusion mistakes
  * Demon/character stat and skill copying to enemies or allies
  * Scripted server side web apps
  * Added verification channel setting and standalone tool for extended content runtime checking
  * Many more accurate field names for client object structures
  * Fixes for some stubborn longtime bugs

 -- COMP Omega <compomega@tutanota.com>  Sat, 30 May 2020 12:40:15 -0600

comphack (4.11.0.1) xenial; urgency=low

  * Corrected PPA build error with tinyxml2.

 -- COMP Omega <compomega@tutanota.com>  Tue, 28 Apr 2020 15:05:13 -0600

comphack (4.11.0) xenial; urgency=low

  * Pentalpha fixes and reward rework (with new settings)
  * Global bazaar cost/time settings
  * Fixing Diaspora base relog bug
  * Fixing double NRA shield use upon target reflect
  * Fixing issue with characters being deleted while in a clan
  * Fixing issues with max speed calculation
  * More missed item/skill restrictions

 -- COMP Omega <compomega@tutanota.com>  Tue, 28 Apr 2020 13:12:06 -0600

comphack (4.10.0) xenial; urgency=low

  * Estoma is now properly broken (as designed)
  * More AI customization
  * AI follow target support
  * Ally NPC creation support for event actions
  * Added skill availability checks for player entities

contrib/cpmaker.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Script to generate data used for conversion between Unicode and Windows code
# pages 932 and 1252.
#
# This file is a contribution to COMP_hack.
#
# Copyright (C) 2012-2020 COMP_hack Team <compomega@tutanota.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from urllib.request import urlopen

import os
import re

# Whether to use normal Unicode mapping or "best fit".
USE_BESTFIT = True

# Generate the URLs to download the Unicode mapping files from.
URL_BASE = "http://www.unicode.org/Public/MAPPINGS/VENDORS/MICSFT/"
URL_CP874 = URL_BASE + ("WindowsBestFit/bestfit874.txt" \
	if USE_BESTFIT else "WINDOWS/CP874.TXT")
URL_CP932 = URL_BASE + ("WindowsBestFit/bestfit932.txt" \
	if USE_BESTFIT else "WINDOWS/CP932.TXT")
URL_CP936 = URL_BASE + ("WindowsBestFit/bestfit936.txt" \
	if USE_BESTFIT else "WINDOWS/CP936.TXT")
URL_CP949 = URL_BASE + ("WindowsBestFit/bestfit949.txt" \
	if USE_BESTFIT else "WINDOWS/CP949.TXT")
URL_CP950 = URL_BASE + "WINDOWS/CP950.TXT"
URL_CP1250 = URL_BASE + ("WindowsBestFit/bestfit1250.txt" \
	if USE_BESTFIT else "WINDOWS/CP1250.TXT")
URL_CP1251 = URL_BASE + ("WindowsBestFit/bestfit1251.txt" \
	if USE_BESTFIT else "WINDOWS/CP1251.TXT")
URL_CP1252 = URL_BASE + ("WindowsBestFit/bestfit1252.txt" \
	if USE_BESTFIT else "WINDOWS/CP1252.TXT")
URL_CP1253 = URL_BASE + ("WindowsBestFit/bestfit1253.txt" \
	if USE_BESTFIT else "WINDOWS/CP1253.TXT")
URL_CP1254 = URL_BASE + ("WindowsBestFit/bestfit1254.txt" \
	if USE_BESTFIT else "WINDOWS/CP1254.TXT")
URL_CP1255 = URL_BASE + ("WindowsBestFit/bestfit1255.txt" \
	if USE_BESTFIT else "WINDOWS/CP1255.TXT")
URL_CP1256 = URL_BASE + ("WindowsBestFit/bestfit1256.txt" \
	if USE_BESTFIT else "WINDOWS/CP1256.TXT")
URL_CP1257 = URL_BASE + ("WindowsBestFit/bestfit1257.txt" \
	if USE_BESTFIT else "WINDOWS/CP1257.TXT")
URL_CP1258 = URL_BASE + ("WindowsBestFit/bestfit1258.txt" \
	if USE_BESTFIT else "WINDOWS/CP1258.TXT")
URL_CP1361 = URL_BASE + "WindowsBestFit/bestfit1361.txt"

# This should appear at the beginning of the file. All instances of %CP% will
# be replaced with the code page number.
FILE_HEADER = """/**
 * @file libcomp/src/LookupTableCP%CP%.h
 * @ingroup libcomp
 *
 * @author COMP Omega <compomega@tutanota.com>
 *
 * @brief Data used for conversion between Unicode and Windows code page %CP%.
 *
 * This file is part of the COMP_hack Library (libcomp).
 *
 * Copyright (C) 2012-2020 COMP_hack Team <compomega@tutanota.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef LIBCOMP_SRC_DECRYPT_LOOKUPTABLECP%CP%_H
#define LIBCOMP_SRC_DECRYPT_LOOKUPTABLECP%CP%_H

/// Table that contains data needed to convert between Unicode
/// and Windows code page %CP%.
static unsigned char LookupTableCP%CP%[] = {
"""

# This should appear at the end of the file. All instances of %CP% will be
# replaced with the code page number.
FILE_FOOTER = """\n};

#endif // LIBCOMP_SRC_DECRYPT_LOOKUPTABLECP%CP%_H\n"""


# Generate a header file that contains an array of mappings between Unicode and
# Windows code page codepoints. The array is simply a series of 16-bit values
# in little endian byte order. For example, it can be cast to the uint16_t type
# and the index would be the Unicode code point. The returned value would then
# be the Windows code page codepoint. Depending on the code page, this value
# may be 8 bits (like CP1252), 16 bits, or either depending on the value of the
# code point (lookup how CP932 works - it is a lot like ShiftJIS). Note that
# this is Unicode code points and not encodings so UTF-8, UTF-16, etc. need to
# be parsed and converted first. To access the other mapping, Windows code page
# to Unicode codepoints, offset the array by 128KiB or 2^16 16-bit elements and
# cast the array to the uint16_t type again. The resulting header files will be
# written to the current working directory.
# url   - URL of the Unicode mapping file to download and use when generating
#         the array of mapping entries.
# cpnum - Number corresponding to the Windows code page to be used. This value
#         is only used for cosmetic things like the name of the file, comments,
#         and the name of the array.
def generateLookup(url, cpnum):
	# Initialize the lookup arrays to zero values. The arrays have 2^16
	# elements to account for all possible values of a 16-bit code point.
	mappingToUnicode = [0 for i in range(2 ** 16)]
	mappingFromUnicode = [0 for i in range(2 ** 16)]

	# Download the file from the specified URL and process each line looking
	# for mapping values.
	if USE_BESTFIT:
		# Formats for different lines in the mapping.
		cpinfoRegExp = re.compile(r"^CPINFO\s+([0-9]+)\s+0x([0-9a-fA-F]+)" +
			r"\s+0x([0-9a-fA-F]+).*$")
		mbtableRegExp = re.compile(r"^MBTABLE\s+([0-9]+).*$")
		wctableRegExp = re.compile(r"^WCTABLE\s+([0-9]+).*$")
		dbcsrangeRegExp = re.compile(r"^DBCSRANGE\s+([0-9]+).*$")
		dbcstableRegExp = re.compile(r"^DBCSTABLE\s+([0-9]+).*$")
		recordRegExp = re.compile(r"^0x([0-9a-fA-F]+)\s+0x([0-9a-fA-F]+).*$")

		# 0 - looking for CODEPAGE
		# 1 - looking for CPINFO
		# 2 - looking for a table
		# 3 - parsing a MBTABLE
		# 4 - parsing a DBCSRANGE
		# 5 - looking for a DBCSTABLE
		# 6 - parsing a DBCSTABLE
		# 7 - parsing a WCTABLE
		parseState = 0

		# How many DBCS lead byte ranges are left to parse.
		rangeCount = 0

		# How many codepoint mappings are left to read from the table.
		recordCount = 0

		# How many lead bytes there are for the DBCS range (each has a
		# matching DBCSTABLE).
		leadByteCount = 0

		# Current lead byte being used.
		leadByte = 0x00

		#
		# For a detailed description of the format, see this page: http://www.
		# unicode.org/Public/MAPPINGS/VENDORS/MICSFT/WindowsBestFit/readme.txt
		#
		for line in urlopen(url).read().decode('utf-8', 'ignore').split('\n'):
			# Debug print the current state and the line being parsed.
			# print str(parseState) + ": " + line

			# If the line is a comment, ignore it.
			if len(line) and line[0] == ';':
				continue

			# Parse the line different depending on the current state.
			if parseState == 0 and line.find("CODEPAGE") == 0:  # Initial state
				# We have found where to start parsing, move to the next state.
				parseState = 1
			elif parseState == 1:  # looking for CPINFO
				# Check if the line matches the CPINFO format.
				match = cpinfoRegExp.match(line)
				if not match:
					continue

				# Default characters during conversion.
				cp = int(match.group(2), 16)
				uni = int(match.group(3), 16)

				# Fill the arrays with default characters.
				mappingToUnicode = [uni for i in range(2 ** 16)]
				mappingFromUnicode = [cp for i in range(2 ** 16)]

				# Move to the next parse state.
				parseState = 2
			elif parseState == 2:  # looking for a table
				match = mbtableRegExp.match(line)
				if match:
					parseState = 3  # parsing a MBTABLE
					recordCount = int(match.group(1))
					continue

				match = dbcsrangeRegExp.match(line)
				if match:
					parseState = 4  # parsing a DBCSRANGE
					rangeCount = int(match.group(1))
					continue

				match = wctableRegExp.match(line)
				if match:
					parseState = 7  # parsing a WCTABLE
					recordCount = int(match.group(1))
					continue

				# If we are at the end stop parsing.
				if line.find("ENDCODEPAGE") == 0:
					break
			elif parseState == 3:  # parsing a MBTABLE record (CP=>Uni)
				match = recordRegExp.match(line)
				if not match:
					continue

				# Get the codepoints for this record.
				cp = int(match.group(1), 16)
				uni = int(match.group(2), 16)

				# Record the conversion.
				mappingToUnicode[cp] = uni

				# Remove the record from the count.
				recordCount -= 1

				# If we have read all records, look for another table.
				if recordCount <= 0:
					parseState = 2
			elif parseState == 4:  # parsing a DBCSRANGE (series of DBCSTABLE)
				match = recordRegExp.match(line)
				if not match:
					continue

				# Determine the first lead byte and the number of lead bytes
				# from the first and last lead byte.
				leadByte = int(match.group(1), 16)
				leadByteCount = int(match.group(2), 16) - leadByte + 1

				# Now that we know how many tables there are, start looking for
				# one.
				parseState = 5
			elif parseState == 5:  # looking for a DBCSTABLE
				match = dbcstableRegExp.match(line)
				if not match:
					continue

				parseState = 6  # parsing a DBCSTABLE
				recordCount = int(match.group(1))
			elif parseState == 6:  # parsing a DBCSTABLE
				match = recordRegExp.match(line)
				if not match:
					continue

				# Get the codepoints for this record.
				# Add the lead byte to the codepoint as the higher 8 bits.
				cp = (leadByte << 8) | int(match.group(1), 16)
				uni = int(match.group(2), 16)

				# Record the conversion.
				mappingToUnicode[cp] = uni

				# Remove the record from the count.
				recordCount -= 1

				# If we have read all records, look for another table.
				if recordCount <= 0:
					# Remove the table from the count.
					leadByteCount -= 1

					# Add to the lead byte.
					leadByte += 1

					# If there is no more lead bytes in the range.
					if leadByteCount <= 0:
						# Remove the range from the count.
						rangeCount -= 1

						# If there is another range to parse, do so; otherwise,
						# look for a new table.
						if rangeCount <= 0:
							parseState = 2  # looking for a table
						else:
							parseState = 4  # parsing a DBCSRANGE
					else:
						parseState = 5  # looking for a DBCSTABLE (next lead)
			elif parseState == 7:  # parsing a WCTABLE record (Uni=>CP)
				match = recordRegExp.match(line)
				if not match:
					continue

				# Get the codepoints for this record.
				cp = int(match.group(2), 16)
				uni = int(match.group(1), 16)

				# Record the conversion.
				mappingFromUnicode[uni] = cp

				# Remove the record from the count.
				recordCount -= 1

				# If we have read all records, look for another table.
				if recordCount <= 0:
					parseState = 2
			else:
				# Should never get here, bail.
				break
	else:
		# This is used to match a line in the mapping file. The format is
		# simpy two hex values with white space in between.
		mappingRegExp = re.compile(r"^0x([0-9a-fA-F]+)\s+0x([0-9a-fA-F]+).*$")

		for line in urlopen(url).read().split('\n'):
			if len(line) and line[0] == '#':
				continue

			# Attempt to match the line to the code point mapping format.
			match = mappingRegExp.match(line)

			# If the line is in the mapping format, process and add it to the
			# lookup arrays.
			if match:
				# The first hex string is the Windows code point and the second is
				# the Unicode code point. Convert both hex strings into integers.
				cp = int(match.group(1), 16)
				uni = int(match.group(2), 16)

				# Add the relationship to the lookup tables.
				mappingToUnicode[cp] = uni
				mappingFromUnicode[uni] = cp

	# Open the output file named LookupTableCP{cpnum}.h where {cpnum} is the
	# code page value passed to the method. After opening the file, write the
	# header replacing all instances of %CP% with the same code page value that
	# was passed into the method. Finally, write a tab to start the first line
	# of the Unicode to Windows code page mapping portion of the array.
	lookupFile = open("LookupTableCP%d.h" % cpnum, "w", newline='\n')
	lookupFile.write(FILE_HEADER.replace("%CP%", str(cpnum)))
	lookupFile.write('\t')

	# Loop over the Unicode to Windows code page mapping array for each of the
	# 2^16 codepoints we intend to convert. Write the two bytes of the matching
	# Windows code page codepoint as two hex bytes in little endian format. If
	# this is the 4th mapping value on a line (which means 8 bytes on the line)
	# then create a new line and indent again; otherwise, add a space.
	for uni in range(2 ** 16):
		cp = mappingFromUnicode[uni]

		lookupFile.write("0x%02x, 0x%02x," % (cp & 0xFF, cp >> 8))

		if (uni % 4) == 3:
			lookupFile.write("\n\t")
		else:
			lookupFile.write(" ")

	# Loop over the Windows code page to Unicode mapping array for each of the
	# 2^16 codepoints we intend to convert. This loop is exactly like the
	# Unicode to Windows code page array above. The conversion code will simply
	# offset 128KiB or 2^16 elements into the array to access this mapping
	# instead of the one above.
	for cp in range(2 ** 16):
		uni = mappingToUnicode[cp]

		lookupFile.write("0x%02x, 0x%02x," % (uni & 0xFF, uni >> 8))

		if (cp % 4) == 3:
			lookupFile.write("\n\t")
		else:
			lookupFile.write(" ")

	# There won't be another line in the array and the last byte should not
	# contain a comma after it so rewind the file by three bytes to remove the
	# characters before writing the footer then write the footer replacing all
	# instances of %CP% with the same code page value that was passed into the
	# method.
	lookupFile.seek(lookupFile.tell() - 3, os.SEEK_SET)
	lookupFile.write(FILE_FOOTER.replace("%CP%", str(cpnum)))

generateLookup(URL_CP874, 874)
generateLookup(URL_CP932, 932)
generateLookup(URL_CP936, 936)
generateLookup(URL_CP949, 949)
generateLookup(URL_CP950, 950)
generateLookup(URL_CP1250, 1250)
generateLookup(URL_CP1251, 1251)
generateLookup(URL_CP1252, 1252)
generateLookup(URL_CP1253, 1253)
generateLookup(URL_CP1254, 1254)
generateLookup(URL_CP1255, 1255)
generateLookup(URL_CP1256, 1256)
generateLookup(URL_CP1257, 1257)
generateLookup(URL_CP1258, 1258)
generateLookup(URL_CP1361, 1361)
